src/Scout/ScoutEngine.php

```diff
@@ -0,0 +1,527 @@
+    {
+        assert(is_string($name), new TypeError(sprintf('Argument #1 ($name) must be of type string, %s given', get_debug_type($name))));
+
+        $this->mongodb->selectCollection($name)->drop();
+    }
+
+    /**
+     * Delete all "search indexes", i.e. all MongoDB collections.
+     */
+    public function deleteAllIndexes()
+    {
+        $collectionNames = $this->mongodb->listCollectionNames([
+            'filter' => [
+                'name' => new Regex('^' . preg_quote($this->prefix)),
+            ],
+        ]);
+
+        foreach ($collectionNames as $collectionName) {
+            $this->mongodb->selectCollection($collectionName)->drop();
+        }
+    }
+
+    /** Get the MongoDB collection used to search for the provided model */
+    private function getSearchableCollection(Model|EloquentCollection $model): MongoDBCollection
+    {
+        if ($model instanceof EloquentCollection) {
+            $model = $model->first();
+        }
+
+        assert(in_array(Searchable::class, class_uses_recursive($model)), sprintf('Model "%s" must use "%s" trait', $model::class, Searchable::class));
+
+        return $this->mongodb->selectCollection($model->searchableAs());
+    }
+
+    /** Get the MongoDB collection used to index the provided model */
+    private function getIndexableCollection(Model|EloquentCollection $model): MongoDBCollection
+    {
+        if ($model instanceof EloquentCollection) {
+            $model = $model->first();
+        }
+
+        assert(in_array(Searchable::class, class_uses_recursive($model)), sprintf('Model "%s" must use "%s" trait', $model::class, Searchable::class));
+
+        return $this->mongodb->selectCollection($model->indexableAs());
+    }
+
+    private static function serialize(mixed $value): mixed
+    {
+        if ($value instanceof DateTimeInterface) {
+            return new UTCDateTime($value);
+        }
+
+        if ($value instanceof Serializable || ! is_iterable($value)) {
+            return $value;
+        }
+
+        // Convert Laravel Collections and other Iterators to arrays
+        if ($value instanceof Traversable) {
+            $value = iterator_to_array($value);
+        }
+
+        // Recursively serialize arrays
+        return array_map(self::serialize(...), $value);
+    }
+
+    private function usesSoftDelete(Model|EloquentCollection $model): bool
+    {
+        if ($model instanceof EloquentCollection) {
+            $model = $model->first();
+        }
+
+        return in_array(SoftDeletes::class, class_uses_recursive($model));
+    }
+
+    private function performSearchIndexOperation(Closure $closure): void
+    {
+        try {
+            $closure();
+        } catch (ServerException $exception) {
+            // @todo add error codes
+            if (in_array($exception->getCode(), [])) {
+                throw new \RuntimeException('Failed to perform search index operation. MongoDB Atlas cluster M10+ is required', 0, $exception);
```

Comment:
https://www.mongodb.com/docs/atlas/atlas-search/shared-tier-limitations/ suggests that Atlas Search is supported by lower tiers. Didn't this come up recently with regard to PHPLIB?